src/wallet/wallet.cpp

```diff
@@ -2294,6 +2294,12 @@ bool CWalletTx::InMempool() const
 }
 
 bool CWalletTx::IsTrusted(interfaces::Chain::Lock& locked_chain) const
+{
+    std::set<uint256> s;
```

Comment:
Hmm I don't know about the need of DoS protection, can't we rely on mempool DEFAULT_ANCESTOR_LIMIT there ? A transaction to be trusted has to be in-mempool (among other standards), so their unconfirmed parent too, at first parent hit which is not we should return. What's th worst complexity for a graph of 25 parents?

Reply:
Even so, you can imagine a case where we have a single input single output chain of 25 taking:

sum n = 0 to N n = n * (n-1) /2 = O(N^2).

Which for 25 = 625 iterations.

Imagine what this would be for a tree with no descendants limit!


The caching guarantees that it is O(N*log(N)) worst case (could be improved possibly by switching to unordered_set... but it's worth benchmarking since we will usually have a very small set)

Reply:
I believe the worst case is O(a^n), where a is the number of inputs/output per transaction, and n is the ancestor limit. Simply build a tx graph where every transaction spends a outputs of the previous layer, in a different inputs.

Reply:
Okay assuming complexity O(a^n), standard txn are limited by MAX_STANDARD_TX_WEIGHT and 148-bytes input, that gives you 548 inputs/outputs on every tx layer so something like 13700 iterations... Would be interesting to know of `IsTrusted` behaves on such computation order, but the fees burn for a DoS attacker seems pretty high.

> Imagine what this would be for a tree with no descendants limit!

Not sure why descendants limit matters there?

Anyway, DoS protection is simple enough to get it for peace of mind.